pinot/generative/torch_gvae/model.py

```diff
@@ -0,0 +1,84 @@
+import torch
+import torch.nn as nn
+import torch.nn.functional as F
+
+from pinot.generative.torch_gvae.layers import GraphConvolution
+
+
+class GCNModelVAE(nn.Module):
+    """Graph convolutional neural networks for VAE
+    """
+    def __init__(self, input_feat_dim, hidden_dim1, hidden_dim2, dropout):
+        """ Construct a VAE with GCN
+        """
+        super(GCNModelVAE, self).__init__()
+        self.gc1 = GraphConvolution(input_feat_dim, hidden_dim1, dropout, act=F.relu)
+        self.dc = InnerProductDecoder(dropout, act=lambda x: x) # Decoder
+        self.output_regression = torch.nn.ModuleList([
+            GraphConvolution(hidden_dim1, hidden_dim2, dropout, act=lambda x: x),
+            GraphConvolution(hidden_dim1, hidden_dim2, dropout, act=lambda x: x),
+        ])
+
+    def forward(self, x, adj):
+        """ Compute the parameters of the approximate Gaussian posterior
+         distribution
+        
+        Args:
+            x (FloatTensor): node features
+                Shape (N, D) where N is the number of nodes in the graph
+            adj (FloatTensor): adjacency matrix
+                Shape (N, N)
+
+        Returns:
+            z: (FloatTensor): latent encodings of the nodes
+                Shape (N, D)
+        """
+        z = self.gc1(x, adj)
+        return z
+
+    def condition(self, x, adj):
+        """ Compute the approximate Normal posterior distribution q(z|x, adj)
+        and associated parameters
+        """
+        z = self.forward(x, adj)
+        theta = [parameter(z, adj) for parameter in self.output_regression]
+        mu  = theta[0]
+        logvar = theta[1]
+        distribution = torch.distributions.normal.Normal(
+                    loc=theta[0],
+                    scale=torch.exp(theta[1]))
+
+        return distribution, theta[0], theta[1]
```

Comment:
Same here. It seems to me that these should already be taken care of if you would just call `pinot.Net`

Reply:
Same point as above, I strongly prefer this to be self-sustaining for a bit until we design the interfaces later on.

Reply:
The thing is that if we wanted to somehow integrate these two to share a representation layer, we'll have to make the **g -> h** part universal. Also that part should be easy to check and rewrite if we were to integrate them later. Now it's a one-line API.

Reply:
Another thing is that hand-write graph convolutional nets (I did a lot before) is a bit error prone and should be avoided. So I highly recommend we use the modular implementation of dgl even if for some reason we shouldn't use the existing one-line API.

Reply:
Okay, so it seems what I can do with the current implementation to compromise is that

the `.forward` function should do what is expected when one calls `representation(g)` within `Net` which is to map from graph to latent representation. And this is the current state of the implementation actually. The `.foward` function maps from `x, adj` to `z` (latent encodings of the nodes).

What I can do is to use DGL implementation instead of handwritten GCN code. This will make it easier to try different DGL architecture in the future.